Automattic/wp-calypso
# Comments: Add new ui.comments.queries state to handle pagination [DEPRECATED]

Unlocks #20737

Add a new UI comments state to handle the pagination requirements of the Comments Management section.

Note: CI failing tests is caused by me not updating unit tests with the new actions changes.
This was my decision to avoid having a scary LOC count. I'll do it once/if we settle with this approach.

## The Problem

We've built the Comments Management section on top of the new comments tree endpoint: fetching _all_ the comments IDs from the server, and use this array of IDs to build and paginate the comments list; then, fetch all the comments details individually.

Unfortunately, this approach turned out too intensive, crippling the section's performances on sites with thousands comments.

This led to reverting to the old (but gold) comments list endpoint, with a more traditional pagination approach of fetching only the page of comments we need, and paginate over the comments count instead of the array of IDs.

Though, Redux doesn't quite like this otherwise tried and tested pagination approach, as it assumes we don't have any gaps in the list of comments.

## The Proposed Solution

Props to @gwwar for this.

This PR creates a new utility state that keeps track of the pages of any possible Comment Management lists:
- Site-wide lists.
- Single post lists.
- Filtered by status.
- Filtered by search query (not implemented here, but the reducer should be easily extendable).

The new state would look like this:
```es6
state.ui.comments.queries = {
  12345678: { // site ID
    site: { // site-wide list
      approved: {
        1: [ 10, 9, 8, 7, 6 ],
        2: [ 5, 4, 3, 2, 1 ],
      },
      'approved?s=foo': { // status and search query
        1: [ 9, 5, 4, 2, 1 ],
      }
    },
    1234: { ... }, // post list
  },
};
```

Each page is populated when the user loads it, in the success handler of the `fetchCommentsList` data layer function.
For example, by opening `comments/all/12345678?page=3`, we'd populate the state like this:
```es6
state.ui.comments.queries[ 12345678 ].site.all = [ /* comments IDs */ ];
```

This creates yet another problem: what happens when one of these page changes?
It can happen after any of these actions:
- We change a comment's status and it moves out of the current filter.
- We delete a comment.
- We reply to a comment, and the reply appears in the same page as the parent comment.

My proposed solution is to let the data layer take care of requesting a fresh copy of the current page in the success handler of all those actions.
Though, for this to work, we need to know all the comments list endpoint query parameters we're currently using.
And, of course, we don't want to call the endpoint when not necessary, but only when we're using the Comments Management section.

So, we create a new `commentsListQuery` object in `CommentList` and we send it down to the actions that would need it.
This means, for example, that the single comment view, which isn't rendered as a `CommentList` child, wouldn't obtain this query object, and therefore will not refresh the page, as it's indeed not necessary for it.

## Testing instructions

A big pain point of this PR is testing it.
As far as I know, there is no way to test it without doing some changes to the code.

Probably the easiest way would be to copy over #20737's changes (https://github.com/Automattic/wp-calypso/pull/20737/files) being careful to not overwrite this PR changes to those files.

Then, try opening `/comments` and navigating around (changing filters, changing pages, going into post view), and see how the `state.ui.comments.queries` state reacts.
